flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessor.java

```diff
@@ -171,10 +183,36 @@ public InputStatus processInput() throws Exception {
 		}
 
 		InputStatus inputStatus = inputProcessors[readingInputIndex].processInput();
+		updatePriorityAvailability();
 		checkFinished(inputStatus, readingInputIndex);
 		return inputSelectionHandler.updateStatus(inputStatus, readingInputIndex);
 	}
 
+	private void updatePriorityAvailability() {
+		if (lastPriorityInputIndex != InputSelection.NONE_AVAILABLE) {
+			final CompletableFuture<?> priorityEventAvailableFuture =
+				inputProcessors[lastPriorityInputIndex].taskInput.getPriorityEventAvailableFuture();
+			// no more priority events for the input
+			if (!priorityEventAvailableFuture.isDone()) {
+				prioritySelectionHandler.setUnavailableInput(lastPriorityInputIndex);
+				if (!prioritySelectionHandler.isAnyInputAvailable()) {
+					priorityAvailability.resetUnavailable();
+				}
+				priorityEventAvailableFuture.thenRun(onPriorityEvent(lastPriorityInputIndex));
+			}
+		}
+	}
+
+	private Runnable onPriorityEvent(int index) {
+		return () -> {
+			// set the priority flag in a mail before notifying StreamTask of availability
+			mainMailboxExecutor.execute(() -> {
+				prioritySelectionHandler.setAvailableInput(index);
+				priorityAvailability.getUnavailableToResetAvailable().complete(null);
+			}, "priority event {}", index);
```

Comment:
nit: add `StreamMultipleInputProcessor` to the mail's name?

Reply:
Commit removed; on `CheckpointedInputGate`, I'm adding the gate.toString().